services/dynamic-sidecar/Dockerfile

```diff
@@ -10,15 +10,24 @@ FROM python:${PYTHON_VERSION}-slim-buster as base
 
 LABEL maintainer="Andrei Neagu <neagu@itis.swiss>"
 
+ARG R_CLONE_VERSION="v1.57.0"
```

Comment:
# @mguidon  @pcrespov @sanderegg writing my proposal here since it stays in the same thread

After adding rclone docker volume plugin, it looked but the following was observed after testing on master:

**PROS:**
- works very well out of the box, with `vfs-cache-policy=minimal` (only writes to the disk when sending the file to the remote), all file access(read and write) is buffered through the disk.
- automatically decides to upload files (which is good to a certain extension)
- file integrity after transferring is kept in place!

**CONS:**
- not working as expected with `vfs-cache-policy=full` (files are first committed to the disk and then synced) reading causes the shamus to get mixed up
- no hooks to detect file sync competition
- very hard to debug, no logs except for some system error
- symlinks were not working (was constraint to change the layout of the jypyter work folder)
- and the final **SHOW STOPPER** very easily you end up in a situation where you cannot remove rclone volumes.  Not even force removal works. This causes very slow `docker volume ls`, taking minutes to finish. The docker daemon needs to be restarted for the node to come back in a working  status (For details: https://github.com/sapk/docker-volume-rclone/issues/59)

As agreed with @mguidon we'll search for an alternative solution.

---

# Proposa 1: rclone in dynamic-sidecar

Using a `watchgod` and `rclone` a `directory_sync` module can be implemented which takes care fo syncing the content to S3.

Pros:
- no difficult problem to do this. I already know who to code it
- way simpler that `rclone docker volume` and `Proposal 2`
- data is first written to the disk and then it is sent to the remote S3 when the file is no longer changed
- multiple invocations of rclone sync will only upload if there is a difference between local and remote

Cons:
- Similar to the current implementation, before opening the service, the directory has to be pulled from S3 via rclone sync

---

# Proposal 2: fuse in user space via python + rclone to sync

Please refer to [this article](https://www.stavros.io/posts/python-fuse-filesystem/) and [this passthrough script](https://github.com/skorokithakis/python-fuse-sample/blob/master/passthrough.py). If you want to test it: remove `allow_other=True`  and run the following command to mirror your `/tmp` folder `mkdir -p tmp_mirror && python passthrough.py.py /tmp $(pwd)/tmp_mirror`

From what I read here we can write a FUSE implementation on top of this passthrough. This would allow us to emulate what the rclone docker volume is doing. Ideally rclone should be used to transfer data.

Pros:
- this will act as the `rclone docker volume`, with the ability for us to tweak and  maintain it.
- the mount can be ran as a task in a different thread in the `dynamic-sidecar`, very easy to debug!

Cons:
- will take longer than above `Proposal 1`, (did not investigate how to connect rclone to the fs primitives)

---

My opinion on the above proposals: 

- `1` is secure and will partially solve the issue faster (we might still end up having to go fo the `2` if we need to mount "file-server" in the study)
- `2` looks achievable (not 100% sure). If it works, on the long run it will be the better alternative.

Reply:
If proposal 2 uses fuse inside the container, the sidecar requires elevated privileges. This is similar to our plan B and we have to think through the security implications. I would prefer to mount than to sync.

Reply:
> > Now that I think about it, [s3fs-fuse](https://github.com/s3fs-fuse/s3fs-fuse) is less flexible that `Proposal 2`, but does the same thing. Also these [Limitations](https://github.com/s3fs-fuse/s3fs-fuse#limitations) apply to both solutions.
> 
> I looked into that but discarded it due the elevated priviliges. Thats why the docker volume plugin is so elegant.

So I was looking on how to implement a docker volume plugin.
This [pyvolume](https://github.com/ronin13/pyvolume) is `python2` and not maintained any longer. **But** we can write a similar one in `python3` and use `s3fs-fuse`.
I think something like this should work. This way you still have a nice plugin which handles thing for you. This should also be ways easier that `Proposal 2`.  I'm not sure how permissions will be handled.

Reply:
some questions: 
1. the link to the docker-r-clone volume repository. Is that thing maintained? the last released is from 2020?
2. Proposal 1: how different is it from now? it looks like exactly the same just the client for upload/download is different? you would also stop zipping the files right?
3. Proposal 2: Did you have a look at the fsspec library? [see here](https://filesystem-spec.readthedocs.io/en/latest/index.html). It is maintained, it's used by the dask guys, and seems to be able to create filesystems. Is also compatible with s3fs-fuse and other backends. Maybe worth a try or analysis?

Reply:
> some questions:
> 
> 1. the link to the docker-r-clone volume repository. Is that thing maintained? the last released is from 2020?
This is the official project linked on the docs of rclone. Last commit in 2021, previous last issue created in sept 2021 and has no replies. We can consider it not actively maintained.

> 2. Proposal 1: how different is it from now? it looks like exactly the same just the client for upload/download is different? you would also stop zipping the files right?
To improve closing times, I would run the sync after file changes as well (this way we can upload the files as soon as they get created in the background). You might end up uploading partial files if in the meanwhile they are being written.
And yes, it uploads files without zipping/unzipping, which also saves some time.

> 3. Proposal 2: Did you have a look at the fsspec library? [see here](https://filesystem-spec.readthedocs.io/en/latest/index.html). It is maintained, it's used by the dask guys, and seems to be able to create filesystems. Is also compatible with s3fs-fuse and other backends. Maybe worth a try or analysis?
Our use case is different. `fsspec` is used to emulate a file system in python, thus only python can access it.
Here a transparent file system which takes care of dealing uploading to s3 is required and other applications like the shell should be able to interact through linux primitives.

Reply:
Case progress and update :

After some more testing we realised that we should remove docker volumes when done with them. This let us to use anonymous  volumes. For details SEE:
-  https://github.com/ITISFoundation/osparc-simcore/pull/2915

Above PR triggered made us stumble upon a different issue with the docker engine. It  creates multiple anonymous volumes when using rclone volume plugin. For details SEE :
- https://github.com/rclone/rclone/issues/6059
 
@mguidon @sanderegg @pcrespov